abi/auto_marshal_abi_spec_test.go

```diff
@@ -0,0 +1,619 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package abi
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"testing"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ava-labs/hypersdk/chain"
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/consts"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+// Combined ABI and AutoMarshal spec
+// Used to verify TypeScript implementation
+// Tests added as needed by TypeScript
+// Ensures consistency in marshaling, not testing Go struct marshaling itself
+
+type AbstractMockAction struct{}
+
+func (AbstractMockAction) ComputeUnits(chain.Rules) uint64 {
+	panic("ComputeUnits unimplemented")
+}
+
+func (AbstractMockAction) Execute(_ context.Context, _ chain.Rules, _ state.Mutable, _ int64, _ codec.Address, _ ids.ID) (outputs [][]byte, err error) {
+	panic("Execute unimplemented")
+}
+
+func (AbstractMockAction) Size() int {
+	// TODO: This has to be automatic for automatic marshalling
+	return 0
+}
+
+func (AbstractMockAction) StateKeys(_ codec.Address, _ ids.ID) state.Keys {
+	panic("StateKeys unimplemented")
+}
+
+func (AbstractMockAction) StateKeysMaxChunks() []uint16 {
+	panic("StateKeysMaxChunks unimplemented")
+}
+
+func (AbstractMockAction) ValidRange(chain.Rules) (start int64, end int64) {
+	panic("ValidRange unimplemented")
+}
+
+type MockActionSingleNumber struct {
+	AbstractMockAction
+	Field1 uint16 `serialize:"true"`
+}
+
+func (MockActionSingleNumber) GetTypeID() uint8 {
+	return 1
+}
+
+type MockActionTransfer struct {
+	AbstractMockAction
+	To    codec.Address `serialize:"true" json:"to"`
+	Value uint64        `serialize:"true" json:"value"`
+	Memo  []byte        `serialize:"true" json:"memo"`
+}
+
+func (MockActionTransfer) GetTypeID() uint8 {
+	return 2
+}
+
+func TestABISpec(t *testing.T) {
```

Comment:
Summary of our offline conversation:

Could we add testing for:

1. `[]abi.TypedField` (doesn't know about the golang type) <-> binary for hash compatibility of the type description
2. Support passing values in via `[]abi.TypedField` so we can add spec tests that the provided values are serialized in the same way as well
3. Split these into shareable test vectors we can re-use across Golang + TypeScript
4. Implement both serialize/deserialize in both Golang + JavaScript so that we can add identity tests as well